packages/vite/src/node/plugins/css.ts

```diff
@@ -2023,3 +2090,162 @@ const preProcessors = Object.freeze({
 function isPreProcessor(lang: any): lang is PreprocessLang {
   return lang && lang in preProcessors
 }
+
+const importLightningCSS = createCachedImport(() => import('lightningcss'))
+
+async function compileLightningCSS(
+  id: string,
+  src: string,
+  config: ResolvedConfig,
+  cssConfig: ResolvedLightningCSSOptions,
+  urlReplacer?: CssUrlReplacer,
+): ReturnType<typeof compileCSS> {
+  const deps = new Set<string>()
+  const res = styleAttrRE.test(id)
+    ? (await importLightningCSS()).transformStyleAttribute({
+        filename: id,
+        code: Buffer.from(src),
+        targets: cssConfig.targets,
+        minify: config.isProduction && config.build.cssMinify,
+        analyzeDependencies: true,
+      })
+    : await (
+        await importLightningCSS()
+      ).bundleAsync({
+        filename: id,
+        resolver: {
+          read(filename) {
+            if (filename === id) {
+              // TODO: add source map inline comment
+              return src
+            }
+            return fs.readFileSync(filename, 'utf-8')
+          },
+          async resolve(id, from) {
+            const publicFile = checkPublicFile(id, config)
+            if (publicFile) {
+              return publicFile
+            }
+
+            const resolved = await getAtImportResolvers(config).css(id, from)
+
+            if (resolved) {
+              deps.add(resolved)
+              return resolved
+            }
+            return id
+          },
+        },
+        targets: cssConfig.targets,
+        minify: config.isProduction && config.build.cssMinify,
+        sourceMap: true,
+        analyzeDependencies: true,
+        cssModules: cssModuleRE.test(id),
+        drafts: { nesting: true },
+      })
+
+  let css = res.code.toString()
+  for (const dep of res.dependencies!) {
+    switch (dep.type) {
+      case 'url':
+        deps.add(dep.url)
+        if (urlReplacer) {
+          css = css.replace(dep.placeholder, await urlReplacer(dep.url, id))
+        }
+        break
+      default:
+        throw new Error(`Unsupported dependency type: ${dep.type}`)
+    }
+  }
```

Comment:
One thing that might be worth trying would be to use the visitor API to do this instead of doing string replacement afterward (using `analyzeDependencies`). You should test the perf and see which is faster. Here's an [example](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Atrue%2C%22nesting%22%3Atrue%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22source%22%3A%22.foo%20%7B%5Cn%20%20background%3A%20url(foo.png)%3B%5Cn%7D%5Cn%22%2C%22visitorEnabled%22%3Atrue%2C%22visitor%22%3A%22%7B%5Cn%20%20Url(url)%20%7B%5Cn%20%20%20%20url.url%20%3D%20'http%3A%2F%2Fexample.com%2F'%20%2B%20url.url%3B%5Cn%20%20%20%20return%20url%3B%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D).

Reply:
I will try it out. A nice thing it that this would give correct source map no?

Reply:
Yes

Reply:
I looked at it but the issue is that the function is async (and it's rooted deep into shared resolve logic being async in Vite)